diff --git a/Functions/Init/.autocomplete__async b/Functions/Init/.autocomplete__async
index e89e041..1e5bd7c 100644
--- a/Functions/Init/.autocomplete__async
+++ b/Functions/Init/.autocomplete__async
@@ -151,14 +151,13 @@ ${0}:precmd() {
 
 .autocomplete:async:clear() {
   # If we exit the ZLE before the last hook widget called, then it won't be able to close the fd.
-  [[
-    -v _autocomplete__async_fd && _autocomplete__async_fd -ge 10 && -t $_autocomplete__async_fd
-  ]] &&
-    exec {_autocomplete_async_fd}<&- &&
+  if [[ -v _autocomplete_async_fd && _autocomplete_async_fd -ge 10 ]]; then
+    builtin zle -F "$_autocomplete_async_fd" 2>/dev/null
+    { exec {_autocomplete_async_fd}<&- } 2>/dev/null
     unset _autocomplete_async_fd
+  fi
   unset curcontext _autocomplete__isearch
 
-  .autocomplete:async:reset-context
   builtin zle -Rc
   return 0
 }
@@ -194,7 +193,7 @@ ${0}:precmd() {
 
   # Unhook ourselves immediately, so we don't get called more than once.
   builtin zle -F "$fd" 2> /dev/null
-  [[ -t $fd ]] && exec {fd}<&-
+  { exec {fd}<&- } 2>/dev/null
 
   if [[ -n $_autocomplete__zle_flags ]]; then
     builtin zle -f $_autocomplete__zle_flags
@@ -370,7 +369,7 @@ ${0}:precmd() {
       (( SECONDS += reply[2] ))
 
     } always {
-      [[ -t $fd ]] && exec {fd}<&-
+      { exec {fd}<&- } 2>/dev/null
     }
 
     [[ -n $curcontext ]] &&
